### PR TITLE
sql/privilege: clean up and increase compatibility of new acl functions

### PR DESCRIPTION
**Pull request by rafiss**
Review note: Please review each commit individually to make this easy to review.
See the Claude Code plan [here](https://gist.github.com/rafiss/9537114859d55b91191b1309ffc6b01c).

---

### sql/privilege: sort ACL characters by privilege bit position 

Replace the hardcoded orderedPrivs list with a sort by Kind value,
which corresponds to the privilege bit position. This produces a
deterministic ACL character ordering and also fixes a pre-existing
bug where `make([]string, len(privileges))` created zero-value
entries before appending.

### sql/privilege: remove redundant ValidACLChars constant 

Use ACLCharToPrivName map lookup instead of maintaining a duplicate
list of valid ACL characters.

### sql: update acldefault to reflect actual default privileges 

Update the acldefault builtin to return privilege defaults that match
what CockroachDB actually assigns to newly created objects:

- Include admin and root roles with ALL privileges, matching
  NewCustomSuperuserPrivilegeDescriptor behavior.
- If the owner is distinct from admin/root, include an owner entry too.
- Adjust privilege chars per object type to match CRDB's defaults
  (e.g. tables get "admrtw" not "arwdDxtm").
- Add parameter ('p') defaults with SET and ALTER SYSTEM privileges.
- Do not mark implicit grant options with '*' in the ACL string.
  PostgreSQL documents that "the owner's implicit grant options are not
  marked in the access privileges display", so we follow suit for
  admin, root, and the owner.

Add TestAclDefaultPublicPrivileges to verify that the public role's
privileges from acldefault match the actual defaults on newly created
objects.

### sql: populate ACL columns in pg_catalog tables 

Populate the relacl, nspacl, datacl, proacl, and typacl columns in
pg_catalog tables using a new privilegeDescriptorToACLArray helper.

Only privileges with a PostgreSQL ACL character equivalent are included;
CockroachDB-specific privileges (BACKUP, CHANGEFEED, ZONECONFIG, etc.)
are skipped.

To match PostgreSQL behavior:

- Return NULL when an object's privileges match its defaults, since
  pg_dump compares ACL columns against acldefault() to decide whether
  to emit GRANT/REVOKE statements.
- Strip grant option '*' markers for admin, root, and the owner,
  since their grant options are implicit (as in PostgreSQL for the
  owner/superuser).
- Update acldefault() to use a shared DefaultACLItems function,
  ensuring consistency between ACL column output and acldefault().

Virtual tables and schemas continue to return NULL ACLs.

### sql/privilege: add ACLItem type for structured aclitem construction 

Add a new ACLItem struct that encapsulates PostgreSQL-compatible aclitem
construction (grantee=privchars/grantor format). This replaces ad-hoc
fmt.Sprintf-based ACL string building with a structured type that:

- Implements redact.SafeFormatter (role names are PII, privilege chars
  are safe)
- Provides NewACLItem constructor and ParseACLItem parser with
  round-trip consistency
- Moves DefaultACLItems to return []ACLItem instead of []string
- Adds IsDefaultACL for order-independent default privilege comparison
- Removes ValidateACLItemString (superseded by ParseACLItem) and the
  now-unused skipACLIdentifier helper

### sql: migrate aclitem callers to use ACLItem type 

Update all callers that previously constructed aclitem strings via
fmt.Sprintf to use the structured ACLItem type:

- NewDACLItemFromDString: use ParseACLItem for validation
- acldefault builtin: use DefaultACLItems ([]ACLItem) for
  descriptor-backed types; use NewACLItem for non-descriptor types
  (except PARAMETER which needs raw ACL chars for SET/ALTER SYSTEM)
- privilegeDescriptorToACLArray: use NewACLItem with ALL expansion
  and IsDefaultACL for default detection
- createDefACLItem: accept username.SQLUsername, use NewACLItem

The aclexplode builtin retains manual char iteration because ACL chars
's' (SET) and 'A' (ALTER SYSTEM) have no privilege.Kind equivalent.

### sql: accept privilege.ACLItem in tree.NewDACLItem 

Change `tree.NewDACLItem` to accept a `privilege.ACLItem` instead of a
string. This eliminates the string→parse→validate round-trip at each
call site where a structured ACLItem is already available.

Callers that still need to construct an aclitem datum from a raw string
(pgwire decoding, casts, random datum generation, string literal
resolution) now use `tree.NewDACLItemFromDString`.

Also change `createDefACLItem` in pg_catalog.go to return
`privilege.ACLItem` instead of `string`, since all callers immediately
pass it to `tree.NewDACLItem`.

---

informs #20296

### CODE REVIEW

File: pkg/sql/pg_catalog.go [hunk 1]
```diff
@@ -1694,27 +1700,111 @@ https://www.postgresql.org/docs/13/catalog-pg-default-acl.html`,
...
+				grantOptions, err = privilege.GetValidPrivilegesForObject(objectType)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		aclItems = append(aclItems, privilege.NewACLItem(
+			grantee, owner, privileges, grantOptions,
+		))
+	}
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/pg_catalog.go:1793]: done

**Review by eric-alton — COMMENTED** [review]
> @eric-alton reviewed 21 files and all commit messages, and made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on fqazi, rafiss, and ZhouXing19).

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAqut-789pK56BeeEt:-OoAqut-789pK56BeeEu:blexik4):*
Outside of the bug fix, the motivation behind this change is unclear. What benefits do we gain from having deterministic ACL ordering?
___
*[`pkg/sql/privilege/kind.go` line 82 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoBBQXd5dnq2ZFFiERH:-OoBBQXd5dnq2ZFFiERI:b-jvipg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/778c8806d2b03dcd4fde14cf356e7d3a215066a4/pkg/sql/privilege/kind.go#L82)):*
> ```Go
> 	// by acldefault, aclexplode, and makeaclitem. They have no
> 	// corresponding CockroachDB privilege and must never be used in
> 	// bitmask operations (Mask/ToBitField/ListFromBitField).
> ```

Can we guard against this in any way? For example, maybe one common code path which performs bitmask operations on these Kind values can check that the value does not exceed `largestKind`.
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 2630 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAwQ0_5IvD_TgdayWQ:-OoAwQ0_5IvD_TgdayWR:b-2bb0ef) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda4a8244153ae589f064fe99d35dba195d39f55/pkg/sql/sem/builtins/pg_builtins.go#L2630)):*
> ```Go
> 				// (matching PostgreSQL behavior for owner defaults).
> 				if def.ownerPrivs != "" {
> 					item := quotedAdmin + "=" + def.ownerPrivs + "/" + quotedOwner
> ```

What is the reason for this ordering, admin before public? (admin, public, root, owner)
___
*[`pkg/sql/pg_catalog.go` line 1748 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAxyAi6VyHFWUfFiKG:-OoAxyAi6VyHFWUfFiKH:b31qsjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/pg_catalog.go#L1748)):*
> ```Go
> // Only privileges that have a PostgreSQL ACL character equivalent are
> // included; CockroachDB-specific privileges (BACKUP, CHANGEFEED, etc.)
> // are silently skipped by ListToACL.
> ```

If our goal is to have pg_dump output CockroachDB-compatible DCL, should these CDB privileges be skipped?
___
*[`pkg/sql/pg_catalog.go` line 1795 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoB2i027wj5FYup1CMU:-OoB2i027wj5FYup1CMV:b-7d4gwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/pg_catalog.go#L1795)):*
> ```Go
> 			return nil, err
> 		}
> 		aclItems = append(aclItems, fmt.Sprintf("%s=%s/%s", grantee, acl, quotedOwner))
> ```

Is the owner being treated as the grantor here? Is that correct, or does it matter?
___
*[`pkg/sql/logictest/testdata/logic_test/pg_catalog` line 5681 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoB5HbnC5M-R6SbZQlL:-OoB5HbnC5M-R6SbZQlM:b5uskfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5681)):*
> ```
> 
> # Verify that ACL columns return NULL when privileges match defaults
> # (matching PostgreSQL behavior), and non-NULL after modification.
> ```

Not all your acl types follow this test pattern. nspacl, datacl, and typacl are all missing a test for when a we modify the default privileges to go from expected NULL to non-NULL output (and back again if the change is reverted).

**Review by rafiss — COMMENTED** [review]
> @rafiss made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on eric-alton, fqazi, and ZhouXing19).

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAqut-789pK56BeeEt:-OoPznqOBQ_f040VOZuX:b-q7lzzd):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

Outside of the bug fix, the motivation behind this change is unclear. What benefits do we gain from having deterministic ACL ordering?
</blockquote></details>

the main benefit is that it makes test output deterministic, so we don't have to deal with frequently rewriting logictests whenever the sort changes due to incidental slice ordering.

customers may also benefit from deterministic ordering too for similar reasons.
___
*[`pkg/sql/pg_catalog.go` line 1748 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAxyAi6VyHFWUfFiKG:-OoQ5rRy7ATzYnZjGDbP:b-b4m7uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/pg_catalog.go#L1748)):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

If our goal is to have pg_dump output CockroachDB-compatible DCL, should these CDB privileges be skipped?
</blockquote></details>

these are getting skipped. but i may have misunderstood you; are you suggesting that they shouldn't be?
___
*[`pkg/sql/pg_catalog.go` line 1795 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoB2i027wj5FYup1CMU:-OoQ5yL09qNutW6isFAq:b-k3a64j) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/pg_catalog.go#L1795)):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

Is the owner being treated as the grantor here? Is that correct, or does it matter?
</blockquote></details>

yeah that is the logic here. CRDB doesn't actually track grantor, so this is a fallback alternative. there's an issue and TODO comment for tracking the grantor elsewhere in this code; i'll add a note here too.
___
*[`pkg/sql/privilege/kind.go` line 82 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoBBQXd5dnq2ZFFiERH:-OoQ6bW_ATrFQUiCJ-nt:b9br787) ([raw file](https://github.com/cockroachdb/cockroach/blob/778c8806d2b03dcd4fde14cf356e7d3a215066a4/pkg/sql/privilege/kind.go#L82)):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

Can we guard against this in any way? For example, maybe one common code path which performs bitmask operations on these Kind values can check that the value does not exceed `largestKind`.
</blockquote></details>

yes, good idea. adding
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 2630 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAwQ0_5IvD_TgdayWQ:-OoQ6prs1uvmpgvhKN7j:b-amouyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda4a8244153ae589f064fe99d35dba195d39f55/pkg/sql/sem/builtins/pg_builtins.go#L2630)):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

What is the reason for this ordering, admin before public? (admin, public, root, owner)
</blockquote></details>

the ordering here is public, admin, root, owner. similar to earlier, it's mostly to make test output deterministic (and since i observed that postgres tends to put the public privileges first)
___
*[`pkg/sql/logictest/testdata/logic_test/pg_catalog` line 5681 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoB5HbnC5M-R6SbZQlL:-OoQ70e75Rymk-vcV6nj:b3k8e60) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5681)):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

Not all your acl types follow this test pattern. nspacl, datacl, and typacl are all missing a test for when a we modify the default privileges to go from expected NULL to non-NULL output (and back again if the change is reverted).
</blockquote></details>

adding the missing test cases

**Review by eric-alton — COMMENTED** [review]
> @eric-alton made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on fqazi, rafiss, and ZhouXing19).

___
*[`pkg/sql/pg_catalog.go` line 1748 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAxyAi6VyHFWUfFiKG:-OoQ8a2b0HaAD0axGC2p:bcp1edo) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/pg_catalog.go#L1748)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

these are getting skipped. but i may have misunderstood you; are you suggesting that they shouldn't be?
</blockquote></details>

I think we may ultimately want to support a pg_dump mode which outputs CDB-compatible DDL/DCL, and a mode which outputs Postgres-compatible DDL/DCL. So... it depends.

**Review by eric-alton — COMMENTED** [review]
> @eric-alton made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on fqazi, rafiss, and ZhouXing19).

___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 2630 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAwQ0_5IvD_TgdayWQ:-OoQEUAVEwlOi1YDAG-Y:bfhtxiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda4a8244153ae589f064fe99d35dba195d39f55/pkg/sql/sem/builtins/pg_builtins.go#L2630)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the ordering here is public, admin, root, owner. similar to earlier, it's mostly to make test output deterministic (and since i observed that postgres tends to put the public privileges first)
</blockquote></details>

I think this may have changed somewhere in your stack of commits. When I made the comment I was looking at admin, public, root, owner as the ordering. Either way, you can resolve this comment now.

**Review by rafiss — COMMENTED** [review]
> @rafiss made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on eric-alton, fqazi, and ZhouXing19).

___
*[`pkg/sql/pg_catalog.go` line 1748 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-OoAxyAi6VyHFWUfFiKG:-OoQFfeX4yrUSDkWnIoj:b-ozoptb) ([raw file](https://github.com/cockroachdb/cockroach/blob/554e8b5248ff7145e6f1d01e50570d9a83c41c19/pkg/sql/pg_catalog.go#L1748)):*
<details><summary><i>Previously, eric-alton (Eric Alton) wrote…</i></summary><blockquote>

I think we may ultimately want to support a pg_dump mode which outputs CDB-compatible DDL/DCL, and a mode which outputs Postgres-compatible DDL/DCL. So... it depends.
</blockquote></details>

i agree. on my draft branch for pg_dump support (https://github.com/cockroachdb/cockroach/pull/166084) there's currently a boolean session variable named `pg_dump_compatibility` that controls if we apply introspection fixups that make cockroachdb output metadata in the way pg_dump expects. before making that draft ready for review, i'll change that to be an enum that can have values `none`, `cockroachdb` or `postgres`.

**Review by rafiss — COMMENTED** [review]
> @rafiss resolved 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on eric-alton, fqazi, and ZhouXing19).

**Review by eric-alton — APPROVED** [review]
:lgtm:

> @eric-alton reviewed 19 files and all commit messages, and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/166185#-:-OoQKzF-76AwsNgIaB4g:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on fqazi and ZhouXing19).
